File: java/Arrays/Day7/Once.java
package Day7;
import java.util.*;

public class Once {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the size of Array: ");
        int n = sc.nextInt();

        int arr[] = new int[n];
        System.out.print("Enter " + n + " " + "numbers: ");

        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        // // brute approach...
        // for (int i = 0; i < n; i++) {
        //     int num = arr[i];
        //     int cnt = 0;

        //     for (int j = 0; j < n; j++) {
        //         if (arr[j] == num) {
        //             cnt++;
        //         }
        //     }

        //     if (cnt == 1) {
        //         System.out.println("Once: " + num);
        //     }
        // }



        // // better approach (using hashing)...
        // int maxi = arr[0];
        // for (int i = 0; i < n; i++) {
        //     maxi = Math.max(maxi, arr[i]);
        // }

        // int hash[] = new int[maxi + 1];

        // for (int i = 0; i < n; i++) {
        //     hash[arr[i]]++;
        // }

        // for (int i = 0; i < n; i++) {
        //     if (hash[arr[i]] == 1) {
        //         System.out.println("Once by better: " + arr[i]);
        //     }
        // }



        // // better approach (using map)...
        // HashMap<Integer, Integer> map = new HashMap<>();

        // for (int i = 0; i < n; i++) {
        //     map.put(arr[i], map.getOrDefault(arr[i], 0) + 1);
        // }

        // for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
        //     if (entry.getValue() == 1) {
        //         System.out.println("Once by better with map: " + entry.getKey());
        //     }
        // }



        // // optimal approach (xor -> only for one unique no.)...
        int XOR = 0;
        for (int i = 0; i < n; i++) {
            XOR ^= arr[i];
        }
    
        System.out.println("Once by optimal: " + XOR);


        sc.close();
    }
}

File: python/Arrays/day7/once.py
n = int(input("Enter the size of array: "))

arr = list(map(int, input(f"Enter {n} numbers: ").split()))

# # brute approach...
# for i in range (n):
#     num = arr[i]
#     cnt = 0

#     for j in range(n):
#         if arr[j] == num:
#             cnt += 1
    
#     if cnt == 1:
#         print("Once by brute is: ", num)



# # better approach (using hashing)...
# maxi = arr[0]

# for i in range(n):
#     maxi = max(maxi, arr[i])

# hash_arr = [0] * (maxi + 1)

# for i in range(n):
#     hash_arr[arr[i]] += 1

# for i in range(n):
#     if hash_arr[arr[i]] == 1:
#         print("Once by better with hash: ", arr[i])




# # better approach (using map)...
# freq_map = {}

# for num in arr:
#     if num in freq_map:
#         freq_map[num] += 1
#     else:
#         freq_map[num] = 1

# for num in arr:
#     if freq_map[num] == 1:
#         print("Once by better with map: ", num)




# # optimal approach (xor -> only for one unique no.)...
XOR = 0
for i in range(n):
    XOR = XOR ^ arr[i]

print("Once by optimal: ", XOR)